admin/server/device_code.go

```diff
@@ -0,0 +1,241 @@
+		return
+	}
+	if authCode.ApprovalState != database.Pending {
+		http.Error(w, "device code already used", http.StatusBadRequest)
+		return
+	}
+	if authCode.Expiry.Before(time.Now()) {
+		http.Error(w, "device code expired", http.StatusBadRequest)
+		return
+	}
+
+	// Update user code with user id and approval
+	authCode.UserID = userID
+	if confirmation != "true" {
+		authCode.ApprovalState = database.Rejected
+	} else {
+		authCode.ApprovalState = database.Approved
+	}
+	err = s.admin.DB.UpdateAuthCode(r.Context(), userCode, userID, authCode.ApprovalState)
+	if err != nil {
+		internalServerError(w, fmt.Errorf("failed to update auth code for userCode: %s, %w", userCode, err))
+	}
+}
+
+// getAccessToken verifies the device code and returns an access token if the request is approved
+func (s *Server) getAccessToken(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		internalServerError(w, fmt.Errorf("failed to read request body: %w", err))
+		return
+	}
+	bodyStr := string(body)
+	values, err := url.ParseQuery(bodyStr)
+	if err != nil {
+		internalServerError(w, fmt.Errorf("failed to parse query: %w", err))
+		return
+	}
+	deviceCode := values.Get("device_code")
+	if deviceCode == "" {
+		http.Error(w, "device_code is required", http.StatusBadRequest)
+		return
+	}
+	grantType := values.Get("grant_type")
+	if grantType != deviceCodeGrantType {
+		http.Error(w, "invalid grant_type", http.StatusBadRequest)
+		return
+	}
+
+	authCode, err := s.admin.DB.FindAuthCodeByDeviceCode(r.Context(), deviceCode)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			http.Error(w, fmt.Sprintf("no such device code: %s found", deviceCode), http.StatusBadRequest)
+			return
+		}
+		internalServerError(w, fmt.Errorf("failed to get auth code for deviceCode: %s, %w", deviceCode, err))
+		return
+	}
+	clientID := values.Get("client_id")
+	if clientID != authCode.ClientID {
+		http.Error(w, "invalid client_id", http.StatusBadRequest)
+		return
+	}
+
+	if authCode.Expiry.Before(time.Now()) {
+		http.Error(w, "expired_token", http.StatusUnauthorized)
+		return
+	}
+	if authCode.ApprovalState == database.Rejected {
+		err = s.admin.DB.DeleteAuthCode(r.Context(), deviceCode)
+		if err != nil {
+			internalServerError(w, fmt.Errorf("failed to clean up rejected code: %s, %w", deviceCode, err))
+			return
+		}
+		http.Error(w, "rejected", http.StatusUnauthorized)
+		return
+	}
+	if authCode.ApprovalState == database.Pending {
+		http.Error(w, "authorization_pending", http.StatusUnauthorized)
+		return
+	}
+	if authCode.ApprovalState != database.Approved || authCode.UserID == "" {
+		internalServerError(w, fmt.Errorf("inconsistent state, %w", errors.New("server error")))
+		return
+	}
+	// TODO handle too many requests
+
+	authToken, err := s.admin.IssueUserAuthToken(r.Context(), authCode.UserID, database.AuthClientIDRillCLI, "CLI login")
```

Comment:
This should use authCode.ClientID, right? And then the CLI client ID (handily hard-coded as `database.AuthClientIDRillCLI`) should be passed by CLI.

I guess the display name has to be empty then. It should be fine, the auth_clients table has display names for clients